### PR TITLE
Add the support of geoToH3 function

### PR DESCRIPTION
**Pull request by yupeng9**
## Description
A function that returns the H3 index address

image

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java [hunk 1]
```diff
@@ -87,7 +87,10 @@
 
   // Geo relationship
   ST_CONTAINS("ST_Contains"),
-  ST_EQUALS("ST_Equals");
+  ST_EQUALS("ST_Equals"),
+
+  // Geo indexing
+  GET_HEXAGON_ADDR("get_hexagon_addr");
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:93]: Suggest renaming it to `"geoToH3"`

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GetHexagonAddressFunction.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  private long[] _results;
+
+  @Override
+  public String getName() {
+    return FUNCTION_NAME;
+  }
+
+  @Override
+  public void init(List<TransformFunction> arguments, Map<String, DataSource> dataSourceMap) {
+    Preconditions
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GetHexagonAddressFunction.java:50]: Should we also support 2 arguments version where the first argument is the serialized points?

File: pinot-core/src/test/java/org/apache/pinot/core/geospatial/transform/GetHexagonAddressFunctionTest.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * under the License.
+ */
+package org.apache.pinot.core.geospatial.transform;
+
+import java.util.Arrays;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.testng.annotations.Test;
+
+
+public class GetHexagonAddressFunctionTest extends GeoFunctionTest {
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/geospatial/transform/GetHexagonAddressFunctionTest.java:26]: (nit) Let's merge it into the `GeoFunctionTest`

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java (hunk 1 above)
Comment (yupeng9) [reply pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:93]: This func name has been used by presto at Uber, so try to be consistent

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GetHexagonAddressFunction.java (hunk 2 above)
Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GetHexagonAddressFunction.java:50]: is it possible to have a different type of first ag? I thought func does not support polymorphism?

File: pinot-core/src/test/java/org/apache/pinot/core/geospatial/transform/GetHexagonAddressFunctionTest.java (hunk 3 above)
Comment (yupeng9) [reply pinot-core/src/test/java/org/apache/pinot/core/geospatial/transform/GetHexagonAddressFunctionTest.java:26]: GeoFunctionTest is the (abstract) base for geo functions.

Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/core/geospatial/transform/GetHexagonAddressFunctionTest.java:26]: I see. You can probably put some javadoc then, and make `GeoFunctionTest` abstract

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GetHexagonAddressFunction.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GetHexagonAddressFunction.java:50]: Yes, we can handle different data type differently in transform function. Check `DistinctCountHLLAggregationFunction` as an example.

Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GetHexagonAddressFunction.java]: updated

**Comment by yupeng9** [comment]
@Jackie-Jiang PTAL. The build failures are not related.

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java [hunk 4]
```diff
@@ -98,4 +99,15 @@ public static String stAsText(byte[] bytes) {
...
   }
+
+  /**
+   * Gets the H3 hexagon address from the location
+   * @param latitude latitude of the location
+   * @param longitude longitude of the location
+   * @param resolution H3 index resolution
+   * @return the H3 index address
+   */
+  public static long getHexagonAddr(double latitude, double longitude, int resolution) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java:110]: Annotate with `ScalarFunction`

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java:110]: We should also add the scalar function with 2 arguments, where the first one is `byte[]` of serialized point

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java:110]: We should put longitude before latitude to be consistent with other geo functions

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:93]: The name should be `getHexagonAddr` for consistency

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GetHexagonAddressFunction.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+import org.apache.pinot.segment.spi.datasource.DataSource;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.locationtech.jts.geom.Geometry;
+
+
+/**
+ *  A function that returns the H3 index address of a given geolocation.
+ */
+public class GetHexagonAddressFunction extends BaseTransformFunction {
+  public static final String FUNCTION_NAME = "GET_HEXAGON_ADDR";
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GetHexagonAddressFunction.java:40]: The name should be `getHexagonAddr` for consistency

**Comment by yupeng9** [comment]
@Jackie-Jiang comments addressed. PTAL

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java [hunk 6]
```diff
@@ -87,7 +87,10 @@
 
   // Geo relationship
   ST_CONTAINS("ST_Contains"),
-  ST_EQUALS("ST_Equals");
+  ST_EQUALS("ST_Equals"),
+
+  // Geo indexing
+  geoToH3("geoToH3");
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:93]: We need all capital case without underscore to make the `getTransformFunctionType()` work
```suggestion
  GEOTOH3("geoToH3");
```

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GeoToH3Function.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+import org.apache.pinot.core.operator.transform.function.TransformFunction;
+import org.apache.pinot.core.plan.DocIdSetPlanNode;
+import org.apache.pinot.segment.local.utils.GeometrySerializer;
+import org.apache.pinot.segment.spi.datasource.DataSource;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.locationtech.jts.geom.Geometry;
+
+
+/**
+ *  A function that returns the H3 index address of a given geolocation.
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GeoToH3Function.java:37]: Put some javadoc on for the supported arguments

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GeoToH3Function.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+ */
+public class GeoToH3Function extends BaseTransformFunction {
+  public static final String FUNCTION_NAME = "geoToH3";
+  private TransformFunction _firstArgument;
+  private TransformFunction _secondArgument;
+  private TransformFunction _thirdArgument;
+  private long[] _results;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/GeoToH3Function.java:44]: You should be able to directly use the `_longValuesSV` within the `BaseTransformFunction`

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java [hunk 9]
```diff
@@ -98,4 +99,16 @@ public static String stAsText(byte[] bytes) {
...
+   * Gets the H3 hexagon address from the location
+   * @param latitude latitude of the location
+   * @param longitude longitude of the location
+   * @param resolution H3 index resolution
+   * @return the H3 index address
+   */
+  @ScalarFunction
+  public static long geoToH3(double longitude, double latitude, int resolution) {
+    return H3Utils.H3_CORE.geoToH3(latitude, longitude, resolution);
+  }
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java:113]: We should also add the version with 2 arguments: `geoToH3(byte[] bytes, int resolution)`

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java (hunk 9 above)
Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java:113]: how does the signature match with same name?

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java:113]: We match function name + argument count, so Pinot can treat them as different function

**Pull request merged**